### PR TITLE
Verify that sizes in `Memory<T>` copy methods are compatible.

### PR DESCRIPTION
**Pull request by YohannDudouit**
The requirements for the `size` parameter is only specified in the documentation for some reason, I think this should be an `MFEM_VERIFY`, or at least an `MFEM_ASSERT`.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#2357](https://github.com/mfem/mfem/pull/2357) | @YohannDudouit | @tzanio | @camierjs + @pazner + @v-dobrev + @acfisher | 06/26/21 | 08/06/21 | 08/08/21 | |

### CODE REVIEW

**Comment by v-dobrev** [comment]
Why remove the documentation?

**Comment by YohannDudouit** [comment]
> Why remove the documentation?

I didn't think it added any value since the contract is enforced by the code, but I can add it back.

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: general/mem_manager.hpp [hunk 1]
```diff
@@ -470,16 +470,22 @@ class Memory
...
 
    /// Copy @a size entries from the host pointer @a src to @a *this.
+   /** The given @a size should not exceed the Capacity() of @a *this. */
    inline void CopyFromHost(const T *src, int size);
 
    /// Copy @a size entries from @a *this to @a dest.
+   /** The given @a size should not exceed the Capacity() of @a *this and the
+       destination, @a dest. */
    inline void CopyTo(Memory &dest, int size) const
    { dest.CopyFrom(*this, size); }
```
Comment (pazner) [inline general/mem_manager.hpp]: Should a check be added here too?

**Review by pazner — COMMENTED** [review]
This looks good and useful to me. The only question is whether the check should be `MFEM_ASSERT` or `MFEM_VERIFY`. Do we have any automated way of identifying and measuring performance regressions?

File: general/mem_manager.hpp (hunk 1 above)
Comment (YohannDudouit) [reply general/mem_manager.hpp]: I missed it, thank you.

**Review by v-dobrev — APPROVED** [review]

**Comment by v-dobrev** [comment]
> The only question is whether the check should be `MFEM_ASSERT` or `MFEM_VERIFY`. Do we have any automated way of identifying and measuring performance regressions?

In these methods, I don't expect performance impact from the checks. The only potential problem with `MFEM_VERIFY` in inline methods is that it duplicates the error handling code which is much larger than the check itself. However, I don't expect these specific methods to be instantiated many times, so using `MFEM_VERIFY` should be fine. That said, I'll be fine with switching to `MFEM_ASSERT` as well.

**Review by pazner — APPROVED** [review]
I agree, the performance implications are almost certainly negligible. I still do think that having automated testing for performance testing would be useful (of course, someone has to find the time to implement it…)

**Comment by tzanio** [comment]
@camierjs and @acfisher, are you interested in reviewing this?

**Comment by tzanio** [comment]
ping: @camierjs and @acfisher

**Review by camierjs — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Pull request merged**